straxen/plugins/event_patternfit.py

```diff
@@ -0,0 +1,435 @@
+            dtype.append((('2LLH per channel for main S2', 's2_2llh_per_channel'),
+                       np.float32, (self.config['n_top_pmts'], )))
+            dtype.append((('2LLH per channel for alternative S2', 'alt_s2_2llh_per_channel'),
+                       np.float32, (self.config['n_top_pmts'], )))
+            dtype.append((('Pattern main S2', 's2_pattern'),
+                       np.float32, (self.config['n_top_pmts'], )))
+            dtype.append((('Pattern alt S2', 'alt_s2_pattern'),
+                       np.float32, (self.config['n_top_pmts'], )))
+            dtype.append((('Pattern for main S1', 's1_pattern'),
+                       np.float32, (self.config['n_tpc_pmts'], )))
+            dtype.append((('2LLH per channel for main S1', 's1_2llh_per_channel'),
+                       np.float32, (self.config['n_tpc_pmts'], )))
+        dtype += strax.time_fields
+        return dtype
+    
+    def setup(self):
+        self.mean_pe_photon = self.config['mean_pe_per_photon']
+        
+        # Getting S1 AFT maps
+        self.s1_aft_map = straxen.InterpolatingMap(
+            straxen.get_resource(self.config['s1_aft_map'], fmt='json'))
+                    
+        # Getting optical maps
+        self.s1_pattern_map = straxen.InterpolatingMap(
+            straxen.get_resource(self.config['s1_optical_map'], fmt='pkl'))
+        self.s2_pattern_map = straxen.InterpolatingMap(
+            straxen.get_resource(self.config['s2_optical_map'], fmt='pkl'))
+        
+        # Getting gain model to get dead PMTs
+        self.to_pe = straxen.get_to_pe(self.run_id, self.config['gain_model'], self.config['n_tpc_pmts'])
+        self.dead_PMTs = np.where(self.to_pe==0)[0]
+        self.pmtbool = ~np.in1d(np.arange(0, self.config['n_tpc_pmts']), self.dead_PMTs)
+        self.pmtbool_top = self.pmtbool[:self.config['n_top_pmts']]
+        self.pmtbool_bottom = self.pmtbool[self.config['n_top_pmts']:self.config['n_tpc_pmts']]
+        
+    def compute(self, events):
+        
+        result = np.zeros(len(events), dtype=self.dtype)
+        result['time'] = events['time']
+        result['endtime'] = strax.endtime(events)
+
+        # Computing LLH values for S1s
+        self.compute_s1_llhvalue(events, result)
+        
+        # Computing LLH values for S2s
+        self.compute_s2_llhvalue(events, result)
+        
+        # Computing binomial test for s1 area fraction top
+        s1_area_fraction_top_probability = np.vectorize(_s1_area_fraction_top_probability)
+        positions = np.vstack([events['x'], events['y'], events['z']]).T
+        aft_prob = self.s1_aft_map(positions)
+        
+        # main s1 events
+        mask_s1 = ~np.isnan(aft_prob)
+        mask_s1 &= ~np.isnan(events['s1_area'])
+        mask_s1 &= ~np.isnan(events['s1_area_fraction_top'])
+        
+        # compute binomial test only if we have events that have valid aft prob, s1 area and s1 aft
+        if len(mask_s1):
```

Comment:
This check is to avoid input length equal to 0 for the vectorized version of _s1_area_fraction_top_probability_ function.

Reply:
should it not be np.sum(mask_s1)? Otherwise if all is False, it will still fail right?

Reply:
You are right, let me update it